### DISCUSSION TITLE
Modelable not working

### DISCUSSION THREAD

**Discussion by vincentdev15** [livewire/livewire#6275] [7. Bugs (archived)]
### Livewire version

3

### Which PHP version are you using?

PHP 8.2

### Steps To Reproduce

I have a parent component (a form) in which I use a child component (a select field).

```
<livewire:components.products-select wire:model.live="form.product_id"></livewire:components.products-select>
```

Here is how I declare the property in the Livewire controller of the child component.

```
#[Modelable] 
public $product_id = null;
```

And here is the view of this component.

```
<select wire:model.live="product_id" class="appearance-none w-full outline-none bg-gray-500 rounded text-white p-3">
    ...
</select>
```

It's the same name, but it could be a different one.

When I change the ```form.product_id``` value from the parent component, the child component (products-select) doesn't select the ```product_id``` updated from the parent component. And it's the same problem when I select a product from the select field, it doesn't update in the parent component.

### Additional information

Nothing else to add ;).

### Please confirm (incomplete submissions will not be addressed)

- [X] I have provided easy and step-by-step instructions to reproduce the bug.
- [X] I have provided code samples as text and NOT images.
- [X] I understand my bug report will be removed if I haven't met the criteria above.

**Comment by reppair** [comment]
Pretty much the same issue here.

However in my case, I don't have a select. Got the following:

```php
// parent component defines
public ?int $playlistId = null;

// parent template renders the child component and specifies the model
<livewire:select-playlist-or-create :$trackId wire:model="playlistId" />

// child component defines
#[Modelable]
public ?int $playlistId = null;

// since I don't have an input, I use a method on the child component to set the prop
public function selectPlaylist(Playlist $playlist): void
{
    $this->playlistId = $playlist->id;
}
```

when I select a playlist trough the child component, then I see the selected playlist id, then I destroy the component from the page (it is in a modal), open it up again, it no longer works... :D

**Reply by reppair** [reply]
Here is something anyone can reproduce in a minute, using the docs example given here: https://livewire.laravel.com/docs/nesting#binding-to-child-data-using-wiremodel
```php
class TodoList extends Component
{
    public $todos = [];

    public $todo = '';

    public function add()
    {
        $this->todos[] = $this->todo;

        $this->reset('todo');
    }
}
```

```html
<div>
    <h1>Todos</h1>

    <livewire:todo-input wire:model="todo" />

    <button wire:click="add">Add Todo</button>

    <div>
        @foreach ($todos as $todo)
            <div>{{ $todo }}</div>
        @endforeach
    </div>
</div>
```

```php
class TodoInput extends Component
{
    #[Modelable]
    public $value = '';
}
```

```html
<div>
    <input type="text" wire:model="value" >
</div>
```

render two or more components.. 
```html
<livewire:todo-list :key="1" />
<livewire:todo-list :key="2" />
<livewire:todo-list :key="3" />
```

1. updating any of the rendered `TodoInput`s instances results in all of them changing their value together on the front-end
2. adding a todo works only for the first component, trying to add a todo in todo lists 2 and 3 does not work.

Similarly, in my case, opening a modal with `TodoList` would work the first time. Even though it is the only instance on the page, after closing the modal and opening it again, the parent no longer "sees" the `value` of the input and can't use it to add a new todo.

**Comment by andybeesee** [comment]
I have a problem that might be related to this, trying to re-use Livewire component multiple times in a form.

The documention says `Currently Livewire only support a single #[Modelable] attribute, only the first one will be bound.` I read this to mean that you could only have one Modelable attribute per component. But I think it actually means you can only have one modelable component on the whole page?

Stripped down example (using Volt for brevity, same thing happens with Livewire components too)

input-one.blade.php
```php
<?php
new class extends Livewire\Volt\Component {
    #[\Livewire\Attributes\Modelable]
    public $textOne;
}
?>
<div>
    <input type="text" wire:model="textOne" placeholder="example" />
</div>
```

input-two.blade.php
```php
<?php
new class extends Livewire\Volt\Component {
    #[\Livewire\Attributes\Modelable]
    public $textTwo;
}
?>
<div>
    <input type="text" wire:model="textTwo" placeholder="example" />
</div>
```

v-one.blade.php
```php
<?php
new class extends Livewire\Volt\Component {
    public $textInputOne = '';
}
?>
<div>
    <livewire:input-one wire:model="textInputOne" />
</div>
```

v-two.blade.php
```php
<?php
new class extends Livewire\Volt\Component {
    public $textInputTwo = '';
}
?>
<div>
    <livewire:input-two wire:model="textInputTwo" />
</div>
```

Then if I try to render those components on a page like...
```blade
 <livewire:v-one />

 <livewire:v-two />
```

I will get an error that says `Method or action [textInputTwo] does not exist on component [v-one].` 

My real-world use case is creating a reusable `user-select` component that would be repeated on a form if they need to select more than one user for something.

Repo with code is here: https://github.com/andybeesee/lw-modelable-issue

**Comment by francoism90** [comment]
Does anyone have a solution? Because I'm facing the same issue. It doesn't update the parents blade/view for some reason. It also doesn't trigger `public function updated()` in the parent.

@andybeesee It doesn't matter, I have tried a single component, and it's still the same issue.

**Comment by lrljoe** [comment]
You currently can't use "live" on the child component binding.

 My hacky workaround is in your child component use updatedValue and push an event to the parent (parent method can just be a return).

**Reply by francoism90** [reply]
Another workaround is to simple use blade components as child, because you can still call the model, right?

It's kinda a hit and miss. I mostly work with traits, so the controller doesn't get bloated, but sometimes adding logic to child's, is useful.

**Reply by vincentdev15** [reply]
Why isn't it possible to use "live" on the child component ? Effectively I have removed "live" and it works. But I'd like to use "live" for some reasons.

**Reply by joshhanley** [reply]
@vincentdev15 it's probably just a bug as modelable is a new feature. Can you submit a failing test PR so we can investigate? Thanks!

**Reply by vincentdev15** [reply]
Now I just wonder if it were not normal that .live doesn't work.

Imagine parent .live fires an event to update the child, then child .live fires an event to update the parent, etc ... it would be an infinite loop.

What do you think about ?

**Reply by lrljoe** [reply]
> @vincentdev15 it's probably just a bug as modelable is a new feature. Can you submit a failing test PR so we can investigate? Thanks!

@joshhanley You can see the lack of live in the features code, lack of live is even referenced in code comments.

Happy to submit a broken test, but from the comments, Caleb is fully aware of it!

**Comment by PhiloNL** [comment]
Thanks for reporting. I've added a failing test https://github.com/livewire/livewire/pull/6762.

**Comment by djaiss** [comment]
I have the same issue.
Modelable doesn't seem to work, even if I reproduce the example in the documentation.

**Reply by francoism90** [reply]
The only 'workaround' is to not use Livewire as child, but use a blade-component instead.

You can still use `wire:model`, and access the parent using `$this`. Not saying it's clean, however you should be really careful nesting Livewire components.

**Comment by rinodrummer** [comment]
I'm having a similar issue: I created nested a component, but the sub-component works dynamically only if I use `wire:model.live` but the source property doesn't get updated.

**Reply by moisish** [reply]
I have the same issue

**Reply by lrljoe** [reply]
The docs do explain the caveat of one modelable property.

What exactly are you experiencing in terms of an error?

Can you share a full example that produces the issue?

**Reply by moisish** [reply]
I am using one modelable property per component as per the documentation.

Parent Component has a public property 
`public array $items = [
        'abc' => [
            'type' => 'text',
            'model' => 'items.abc.data',
            'data' => []
        ],
        'efg' => [
            'type' => 'date',
            'model' => 'items.efg.data',
            'data' => []
        ],
    ];`

and I have nested child components for each type

The "text" child component has a public modelable property 
`
    #[Modelable]
    public array $data = ['text' => ''];
`

the modelable from the child component does not update the parent component

**Reply by moisish** [reply]
My other option is to use events and dispatch the changes from child to parent but this will only work if the child component is using wire:model.blur/live - not being lazy.

What I am trying to achieve is having the form submit on the parent component and gather all the data from the child components when the form gets submitted

**Reply by moisish** [reply]
@PhiloNL @joshhanley can you share some
light here please?

**Reply by francoism90** [reply]
@moisish Are you writing a wizard?

You can create a global state on your first/parent Livewire-component, and push the data of the child to it's parent.

The child-components can be just Blade Components, and you can dispatch/wire:model, directly using your child blade.

I try to avoid Livewire-inside-Livewire components, as the docs already explain they are basically islands/not connected in anyway. Since they inherit, you can also call `$this->variableName` inside your Blade.

**Reply by joshhanley** [reply]
@moisish Can you make a https://wirebox.app showing your problem so we can have a look?

**Reply by lrljoe** [reply]
> What I am trying to achieve is having the form submit on the parent component and gather all the data from the child components when the form gets submitted

It really sounds like what you're trying to do, should really be accomplished with child View Components, **not** child Livewire Components.
Using your example, in your **Parent** Livewire Component, item you're defining a list of items.  For example:
'abc' has:
```
'type' => 'text'
'model' => 'items.abc.data'
'data' => []
```
Given this, I'm assuming that ultimately, you're looking to bind a text "input" to "items.abc.data"?

If I'm understanding your example correctly, why aren't you just using a blade/view component to display the input?  It can of course be an anonymous component.  This would negate the need to use Modelable entirely, as it's directly wire:model'd into your parent component.

**Personally** - I'd only be using the "Modelable" approach when I'm using something like a Select2 input, where I want to have an input, that is making calls to the backend "live" independently of other elements.  i.e. where I don't want to cause lifecycles/interactions with the **Parent** or Sibling components.
Otherwise, you're abstracting to a Livewire Component needlessly, which the docs strongly recommend against!

As @joshhanley mentioned, if you can give a wirebox example, then this helps massively in diagnosing any issues.

**Comment by joydeep-bhowmik** [comment]
is this not fixed yet ?

**Reply by lrljoe** [reply]
I mean, Modelable works, what's the specific issue you're encountering?

**Reply by zircuitz** [reply]
I'm also having issues with this. 

When I change the modelable property on a child component, it doesn't update the parent component for some reason.

If I push it through an event however, it works fine.

**Reply by lrljoe** [reply]
> I'm also having issues with this. 
> 
> When I change the modelable property on a child component, it doesn't update the parent component for some reason.
> 
> If I push it through an event however, it works fine.

Can you share your code please?

Mainly
The property in the Parent Component

The section  in the Parent Component's view/blade referencing the Child Component

The property in the Child Component

The relevant view/blade sections for your Child Component

**Comment by reppair** [comment]
That should've been converted to an issue by now. Eventually even fixed or at least explained WHY it is not working as we would expect it to. See this again - https://github.com/livewire/livewire/discussions/6275#discussioncomment-6784562

**Reply by lrljoe** [reply]
Please can you give an up-to-date example, of what you're doing, as modelable 100% works when used as documented.

**Reply by francoism90** [reply]
> Currently Livewire only supports a single #[Modelable] attribute, so only the first one will be bound

I don't think the Todo example works, because it binds them on multiple items. Better is to pass todos as modelable and refresh the view when you remove one.

I know this isn't great, but it's indeed broken for quite some time.

**Reply by reppair** [reply]
So that -> https://github.com/livewire/livewire/discussions/6275#discussioncomment-6784562 <- is not the way it is meant to work and is not the documented way?

**Reply by lrljoe** [reply]
@reppair - I asked you for an example of what you're doing, that isn't working.  Sharing threads from a year ago, isn't going to engage anyone to provide you with assistance.

Without you providing any additional information, my only guess is that you're doing something that isn't supported.

Sharing code examples allows people to replicate, diagnose the issue, and provide assistance.  Without that, everyone's just guessing, and nobody has time in their day for that!

**Reply by reppair** [reply]
@lrljoe the thread I've created a year ago and I'm linking to, is literally the example from the documentation and what is not not working for me. Nor it is for others. So, if you think you could help us understand why it is not working, that would be great.

If we can't make a simple example snippet work as intended and documented, what is the point of discussing my personal use case? It would only bloat the thread really. It is exactly why I picked up a simple example, from the docs (at the time) - to ensure it is not something specific to my use case and anyone can  reproduce in 5 min.

**Reply by lrljoe** [reply]
@reppair  - Taking the opposite of what you've said.  I've got Modelable readily working across dozens of components.  If you can share a simple example, based on how you've read the documentation, how you've implemented it, then it'll allow me to understand what's going on.

I can then advise you how to fix it, and I can get the documentation updated if it's proving unclear.

It's really difficult when you know the framework like the back of your hand to understand where the documentation is lacking, as when I read it, I know what it's trying to explain, but I appreciate that might not be the case for everyone!

I can see why it **might** be confusing from the docs example, so I just need to know how you've interpreted it, so that I can effectively update the docs.

The example DOES work, but it does :
- admittedly seem to assume some pre-existing knowledge, so I'm keen to understand how you interpreted it, otherwise any update I do isn't necessarily going to help you out!
- Not provide a simple, straight-forward, replicable example, as the docs do jump around a bit

**Reply by lrljoe** [reply]
[https://wirebox.app/b/y8392](https://wirebox.app/b/y8392) here's a very basic example of a Modelable working

[https://wirebox.app/b/yr505](https://wirebox.app/b/yr505) The same example. with Computed as a concept with the same base example:.

[https://wirebox.app/b/4je88](https://wirebox.app/b/4je88) The same example, using some Events/On

**Reply by reppair** [reply]
Hey @lrljoe here is a wirebox example https://wirebox.app/b/xvgqe. Apparently it is now working!

It's been fixed in some version that past year. 🎉 

Unfortunately my local example is no longer to be found so I am not sure which version of LIvewire 3 I had at the time (whatever version was the latest at the time - Aug 21, 2023).

Not important, the fact is, with the latest version it is working and I can go re-open the ticket I have and replace the events with modelable attributes now!

Thank you for the effort!

**Comment by lrljoe** [comment]
@vincentdev15  - please mark this as answered!
